subprojects/core/src/test/groovy/org/gradle/api/internal/attributes/DefaultMutableAttributeContainerTest.groovy

```diff
@@ -30,6 +30,71 @@ import spock.lang.Specification
 class DefaultMutableAttributeContainerTest extends Specification {
     def attributesFactory = AttributeTestUtil.attributesFactory()
 
+    def "lazy attributes are evaluated in insertion order"() {
+        def container = new DefaultMutableAttributeContainer(attributesFactory)
+        def actual = []
+        def expected = []
+        (1..100).each { idx ->
+            def testAttribute = Attribute.of("test"+idx, String)
+            expected << idx
+            container.attributeProvider(testAttribute, Providers.<String>changing {
+                actual << idx
+                "value " + idx
+            })
+        }
+        expect:
+        container.asImmutable().keySet().size() == 100
+        actual == expected
+    }
+
+    def "realizing the value of lazy attributes may cause other attributes to be realized"() {
+        def container = new DefaultMutableAttributeContainer(attributesFactory)
+        def firstAttribute = Attribute.of("first", String)
+        def secondAttribute = Attribute.of("second", String)
+        container.attributeProvider(firstAttribute, Providers.<String>changing {
+            // side effect is to evaluate the secondAttribute's value and prevent
+            // it from changing by removing it from the list of "lazy attributes"
+            container.getAttribute(secondAttribute)
+            "first"
+        })
+        container.attributeProvider(secondAttribute, Providers.of("second"))
+
+        expect:
+        container.asImmutable().keySet() == [secondAttribute, firstAttribute] as Set
```

Comment:
This actually seems wrong, but it's current behavior. 

The order of attributes out of the container is insertion order for eager values and _realization_ order for lazy values. 

I think to fix this properly, we would probably get rid of the distinction between eager/lazy and just store everything as lazy so we don't have to do the insertion juggling.

Reply:
This was an idea I had when working on lazy Attributes, but I worried it would be an unacceptable performance regression to do this for all Attributes.  It would certainly make the attribute container code cleaner.